dockertest/docker_daemon.py

```diff
@@ -112,51 +111,136 @@ def version(self):
+    """
+    return '.docker-autotest-preserved'
+
 
-    :param daemon_process: Opaque daemon_process object (not for direct use)
-    :param regex: Regular expression to search for
-    :param timeout: Maximum time to wait before returning False
+def assert_pristine_environment():
     """
-    return wait_for_output(daemon_process.get_stderr,
-                           regex,
-                           timeout=timeout)
+    Barf if there are any leftover .docker-autotest-preserved files
+    in /etc/sysconfig; this would indicate that a previous test
+    failed to clean up properly, and our system is in an
+    undefined state.
+    """
+    for suffix in ['', '-latest']:
+        path = '/etc/sysconfig/docker%s%s' % (suffix, _preserved_extension())
+        if os.path.exists(path):
+            raise RuntimeError("Leftover backup file: %s. System is"
+                               " in undefined state! Please examine that"
+                               " and its original file; if appropriate,"
+                               " mv it back into place." % path)
+
+
+def edit_options_file(remove=None, add=None):
+    """
+    Write a new /etc/sysconfig/docker* file with new OPTIONS string.
+    Preserve the original.
 
+    :param remove: string or list of strings - option(s) to remove from line
+    :param add: string or list of strings - option(s) to add to line
+    """
+    sysconfig_file = '/etc/sysconfig/%s' % _which_docker()
+    sysconfig_bkp = sysconfig_file + _preserved_extension()
+    if os.path.exists(sysconfig_bkp):
+        raise RuntimeError("Backup file already exists: %s" % sysconfig_bkp)
+    sysconfig_tmp = sysconfig_file + '.tmp'
+    with open(sysconfig_file, 'r') as sysconfig_fh_in:
+        with open(sysconfig_tmp, 'w') as sysconfig_fh_out:
+            for line in sysconfig_fh_in:
+                if line.startswith('OPTIONS='):
+                    line = edit_options_string(line, remove=remove, add=add)
+                sysconfig_fh_out.write(line)
+    os.link(sysconfig_file, sysconfig_bkp)
+    os.rename(sysconfig_tmp, sysconfig_file)
+
+
+def edit_options_string(line, remove=None, add=None):
+    """
+    Helper for edit_options_file(). Given an OPTIONS='...' string,
+    returns OPTIONS='...'\n with the given options removed and/or added.
 
-def restart_service(daemon_process=None):
+    :param line: string of the form OPTIONS='something'
+    :param remove: string or list of strings - option(s) to remove from line
+    :param add: string or list of strings - option(s) to add to line
     """
-    Restart the docker service using host OS's service manager
+    if not line.startswith('OPTIONS='):
+        raise ValueError("input line does not start with OPTIONS= : %s" % line)
+    line = line[8:].rstrip()
+    quote = ''
+    if line[0] == '"' or line[0] == "'":
+        quote = line[0]
+        if line[-1] != quote:
+            raise ValueError("mismatched quotes in %s" % line)
+        line = line[1:-1]
```

Comment:
Oh, maybe it's my mistake misunderstanding this method.  I was thinking something like:

```python
def f(line,adds=[],removes=[]):
    new_ops = []
    ops = shlex.split(line, False)
    while ops:
        _op = ops.pop()
        if _op not in removes:
            new_ops.append('"%s"' % _op)
    try:
        while adds:
            new_ops.append(adds.pop())
    except AttributeError:
            new_ops.append(adds)
    new_ops.reverse()
    return "OPTIONS='%s'" % ' '.join(new_ops)
```

Funky looking output, but I think it will always be shell-correct.  Anyway, it's only 4 lines shorter than what you have, so I suppose it doesn't much matter.